# Used Car Price Prediction

In [2]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_csv("quikr_car.csv")
df

name    company  year          Price  \
0      Hyundai Santro Xing XO eRLX Euro III    Hyundai  2007         80,000   
1                   Mahindra Jeep CL550 MDI   Mahindra  2006       4,25,000   
2                Maruti Suzuki Alto 800 Vxi     Maruti  2018  Ask For Price   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT    Hyundai  2014       3,25,000   
4          Ford EcoSport Titanium 1.5L TDCi       Ford  2014       5,75,000   
..                                      ...        ...   ...            ...   
887                                      Ta       Tara  zest       3,10,000   
888                     Tata Zest XM Diesel       Tata  2018       2,60,000   
889                      Mahindra Quanto C8   Mahindra  2013       3,90,000   
890                Honda Amaze 1.2 E i VTEC      Honda  2014       1,80,000   
891               Chevrolet Sail 1.2 LT ABS  Chevrolet  2014       1,60,000   

     kms_driven fuel_type  
0    45,000 kms    Petrol  
1        40 kms    Diesel  
2    22,000 kms    Petrol  
3    28,000 kms    Petrol  
4    36,000 kms    Diesel  
..          ...       ...  
887         NaN       NaN  
888  27,000 kms    Diesel  
889  40,000 kms    Diesel  
890      Petrol       NaN  
891      Petrol       NaN  

[892 rows x 6 columns]

In [4]:
data_copy = df.copy()

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 41.9+ KB


In [6]:
df.isna().sum()

name           0
company        0
year           0
Price          0
kms_driven    52
fuel_type     55
dtype: int64

In [7]:
df["company"].unique()

array(['Hyundai', 'Mahindra', 'Maruti', 'Ford', 'Skoda', 'Audi', 'Toyota',
       'Renault', 'Honda', 'Datsun', 'Mitsubishi', 'Tata', 'Volkswagen',
       'I', 'Chevrolet', 'Mini', 'BMW', 'Nissan', 'Hindustan', 'Fiat',
       'Commercial', 'MARUTI', 'Force', 'Mercedes', 'Land', 'Yamaha',
       'selling', 'URJENT', 'Swift', 'Used', 'Jaguar', 'Jeep', 'tata',
       'Sale', 'very', 'Volvo', 'i', '2012', 'Well', 'all', '7', '9',
       'scratch', 'urgent', 'sell', 'TATA', 'Any', 'Tara'], dtype=object)

In [8]:
df["fuel_type"].unique()

array(['Petrol', 'Diesel', nan, 'LPG'], dtype=object)

## Data Cleaning

## Price

In [9]:
l = []
for i in range(len(df)):
    try:
        k = df["Price"][i].split(",")
        s = int("".join(k))
        l.append(s)
    except:
        l.append(None)
print(l.count(None))

35


In [10]:
df["Price"] = l

#### Deleting all the rows whose ```Price``` value is ```NaN``` or ```None```

In [11]:
df.drop(df[df["Price"].isna()].index,inplace=True)

In [12]:
df["Price"] = df["Price"].astype(int)

## fuel_type

In [13]:
df["fuel_type"].value_counts()

Petrol    428
Diesel    386
LPG         2
Name: fuel_type, dtype: int64

In [14]:
df["fuel_type"].isna().sum()*100/len(df["fuel_type"])

4.784130688448075

#### Since the data missing is 4.78% of the remainging data we can remove the rows with na values

In [15]:
df.drop(df[df["fuel_type"].isna()].index,inplace=True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 816 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        816 non-null    object
 1   company     816 non-null    object
 2   year        816 non-null    object
 3   Price       816 non-null    int32 
 4   kms_driven  816 non-null    object
 5   fuel_type   816 non-null    object
dtypes: int32(1), object(5)
memory usage: 41.4+ KB


## year

In [17]:
df["year"].unique()

array(['2007', '2006', '2014', '2012', '2013', '2016', '2015', '2010',
       '2017', '2008', '2018', '2011', '2019', '2009', '2005', '2000',
       '2003', '2004', '1995', '2002', '2001'], dtype=object)

In [18]:
df["year"].isna().sum()*100/len(df["year"])

0.0

## kms_driven

In [19]:
df["kms_driven"].isna().sum()

0

In [20]:
df["kms_driven"]

0        45,000 kms
1            40 kms
3        28,000 kms
4        36,000 kms
6        41,000 kms
           ...     
883      50,000 kms
885      30,000 kms
886    1,32,000 kms
888      27,000 kms
889      40,000 kms
Name: kms_driven, Length: 816, dtype: object

### converting obj to ```int``` and making all other values ```None``` which hold other values

In [21]:
import re

In [22]:
ll = []
sep = "[, ]"
for i in range(len(df)):
    try:
        k = re.split(sep,df["kms_driven"][i])
        k.pop()  # removing kms
        s = int("".join(k))
        ll.append(s)
        # print(s)
    except:
        ll.append(None)
print(ll.count(None))

65


In [23]:
print(ll.count(None)*100/len(ll))

7.965686274509804


In [24]:
df["kms_driven"] = ll

In [25]:
df.drop(df[df["kms_driven"].isna()].index,inplace=True)
df

name   company  year   Price  kms_driven  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000     45000.0   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006  425000        40.0   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000     28000.0   
6                               Ford Figo      Ford  2012  175000     36000.0   
8        Ford EcoSport Ambiente 1.5L TDCi      Ford  2016  830000     41000.0   
..                                    ...       ...   ...     ...         ...   
865                   Honda City 1.5 S MT     Honda  2009  270000     20000.0   
875                  Honda City VX Petrol     Honda  2016  860000    100000.0   
886                  Toyota Corolla Altis    Toyota  2009  300000      7000.0   
888                   Tata Zest XM Diesel      Tata  2018  260000     31000.0   
889                    Mahindra Quanto C8  Mahindra  2013  390000     33000.0   

    fuel_type  
0      Petrol  
1      Diesel  
4      Diesel  
6      Diesel  
8      Diesel  
..        ...  
865    Petrol  
875    Petrol  
886    Petrol  
888    Diesel  
889    Diesel  

[751 rows x 6 columns]

In [26]:
df["kms_driven"] = df["kms_driven"].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 751 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        751 non-null    object
 1   company     751 non-null    object
 2   year        751 non-null    object
 3   Price       751 non-null    int32 
 4   kms_driven  751 non-null    int32 
 5   fuel_type   751 non-null    object
dtypes: int32(2), object(4)
memory usage: 35.2+ KB


## company

In [40]:
df["company"].value_counts()

Maruti        201
Hyundai       126
Mahindra       91
Tata           59
Honda          58
Toyota         34
Renault        33
Chevrolet      29
Ford           28
Volkswagen     15
Skoda          12
Audi           10
BMW             8
Datsun          7
Mini            7
Mitsubishi      6
Nissan          6
Mercedes        5
Force           4
Hindustan       3
Fiat            3
Jaguar          2
Land            1
Jeep            1
Volvo           1
Name: company, dtype: int64

In [28]:
df["company"].isna().sum()

0

## name

### on observing ```name``` column, there are many unnecessary information like model number and type which is not that useful.
### we will be using first 2 to 3 words in the name

In [29]:
df["name"]

0      Hyundai Santro Xing XO eRLX Euro III
1                   Mahindra Jeep CL550 MDI
4          Ford EcoSport Titanium 1.5L TDCi
6                                 Ford Figo
8          Ford EcoSport Ambiente 1.5L TDCi
                       ...                 
865                     Honda City 1.5 S MT
875                    Honda City VX Petrol
886                    Toyota Corolla Altis
888                     Tata Zest XM Diesel
889                      Mahindra Quanto C8
Name: name, Length: 751, dtype: object

In [30]:
l = []

for i in df["name"]:
    l.append(" ".join(i.split()[:3]))

for i in l[:4]:
    print(i)

Hyundai Santro Xing
Mahindra Jeep CL550
Ford EcoSport Titanium
Ford Figo


In [31]:
df["name"] = l

In [32]:
df

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
4    Ford EcoSport Titanium      Ford  2014  575000       28000    Diesel
6                 Ford Figo      Ford  2012  175000       36000    Diesel
8    Ford EcoSport Ambiente      Ford  2016  830000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
865          Honda City 1.5     Honda  2009  270000       20000    Petrol
875           Honda City VX     Honda  2016  860000      100000    Petrol
886    Toyota Corolla Altis    Toyota  2009  300000        7000    Petrol
888            Tata Zest XM      Tata  2018  260000       31000    Diesel
889      Mahindra Quanto C8  Mahindra  2013  390000       33000    Diesel

[751 rows x 6 columns]

In [33]:
df.describe()

Price     kms_driven
count  7.510000e+02     751.000000
mean   4.189434e+05   45708.210386
std    4.860138e+05   34031.413036
min    3.000000e+04       0.000000
25%    1.790000e+05   27000.000000
50%    3.000000e+05   41000.000000
75%    4.999990e+05   55900.000000
max    8.500003e+06  400000.000000

### It can be seen that there is an outlier in our data because the difference btw 75th percentile and the max value is quite significant.

In [34]:
df[df["Price"]>6000000]

name   company  year    Price  kms_driven fuel_type
562  Mahindra XUV500 W6  Mahindra  2014  8500003       72160    Diesel

In [35]:
df = df[df["Price"]<6000000]

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 750 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        750 non-null    object
 1   company     750 non-null    object
 2   year        750 non-null    object
 3   Price       750 non-null    int32 
 4   kms_driven  750 non-null    int32 
 5   fuel_type   750 non-null    object
dtypes: int32(2), object(4)
memory usage: 35.2+ KB


## After cleaning our data the remaining number of rows are 750 out of 892 i.e. 15.91% of data loss.

In [77]:
df.to_csv("cleaned_cars_data.csv")

## Model

### Since we want to predict the prices of the used cars, this problem is a regression problem and our target column is Price and rest are independent features.

In [37]:
X = df.drop(columns=["Price"])
y = df["Price"]

In [45]:
X

name   company  year  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006          40    Diesel
4    Ford EcoSport Titanium      Ford  2014       28000    Diesel
6                 Ford Figo      Ford  2012       36000    Diesel
8    Ford EcoSport Ambiente      Ford  2016       41000    Diesel
..                      ...       ...   ...         ...       ...
865          Honda City 1.5     Honda  2009       20000    Petrol
875           Honda City VX     Honda  2016      100000    Petrol
886    Toyota Corolla Altis    Toyota  2009        7000    Petrol
888            Tata Zest XM      Tata  2018       31000    Diesel
889      Mahindra Quanto C8  Mahindra  2013       33000    Diesel

[750 rows x 5 columns]

In [43]:
y

0       80000
1      425000
4      575000
6      175000
8      830000
        ...  
865    270000
875    860000
886    300000
888    260000
889    390000
Name: Price, Length: 750, dtype: int32

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 750 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        750 non-null    object
 1   company     750 non-null    object
 2   year        750 non-null    object
 3   Price       750 non-null    int32 
 4   kms_driven  750 non-null    int32 
 5   fuel_type   750 non-null    object
dtypes: int32(2), object(4)
memory usage: 35.2+ KB


In [50]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
import pickle

In [48]:
categorical_cols = ["name", "company", "year", "fuel_type"]

In [57]:
ohe = OneHotEncoder()
ohe.fit(X[categorical_cols])

OneHotEncoder()

In [58]:
ohe.categories_

[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A8',
        'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
        'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
        'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
        'Chevrolet Beat LS', 'Chevrolet Beat LT', 'Chevrolet Beat PS',
        'Chevrolet Cruze LTZ', 'Chevrolet Enjoy', 'Chevrolet Enjoy 1.4',
        'Chevrolet Sail 1.2', 'Chevrolet Sail UVA', 'Chevrolet Spark',
        'Chevrolet Spark 1.0', 'Chevrolet Spark LS', 'Chevrolet Spark LT',
        'Chevrolet Tavera LS', 'Chevrolet Tavera Neo', 'Datsun GO T',
        'Datsun Go Plus', 'Datsun Redi GO', 'Fiat Petra ELX',
        'Fiat Punto Emotion', 'Force Motors Force', 'Force Motors One',
        'Ford EcoSport', 'Ford EcoSport Ambiente',
        'Ford EcoSport Titanium', 'Ford EcoSport Trend',
        'Ford Endeavor 4x4', 'Ford Fiesta', 'Ford Fiesta SXi', 'Ford Figo',
        'Ford Figo Diesel', 'Ford Figo Durator

In [74]:
pipeline = Pipeline([
    ('preprocessor', ColumnTransformer(
        transformers=[
            ('cat', OneHotEncoder(categories=ohe.categories_,drop='if_binary'), categorical_cols)],
        remainder='passthrough')),
    ('regressor', LinearRegression())
])

In [79]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=12)


In [80]:
pipeline.fit(X_train,y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A8',
       'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
       'Chevrolet Beat LS', '...
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['1995', '2000', '2001', '2002', '2003', '2004', '2005', '2006',
       '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014',
       '2015', '2016', '2017', '2018', '2019'], dtype=object),
                                                                            array(['Diesel', 'LPG', 'Petrol'], dtype=object)],
                                                                drop='if_binary'),
                                                  ['name', 'company', 'year',
                                                   'fuel_type'])])),
                ('regressor', LinearRegression())])

In [81]:
y_pred = pipeline.predict(X_test)

In [82]:
r2_score(y_test, y_pred)

0.7039644946357699

### this less r2 score is beacuse of the less amount of data available.
#### we can put this in a loop untill we get the max r2 score, as it is prossible the random sampling of test and train split could lead to the less r2 score.

In [91]:
l = []
for i in range(200):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=i)
    pipeline.fit(X_train,y_train)
    y_pred = pipeline.predict(X_test)
    l.append((r2_score(y_test, y_pred),i))


In [92]:
l.sort(reverse=True)
print(l[0])

(0.8383633500286412, 62)


### r2 score for 62th random state is maximum so we will be training our model with that train text split

In [93]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=62)
pipeline.fit(X_train,y_train)
y_pred = pipeline.predict(X_test)

In [94]:
r2_score(y_test, y_pred)

0.8383633500286412

In [95]:
with open('linear_regression_model.pkl', 'wb') as file:
    pickle.dump(pipeline, file)

In [114]:
# pipeline.predict(pd.DataFrame([["Maruti Suzuki Swift","Maruti","2019",100000,"Petrol"]],columns=["name","company","year","kms_driven","fuel_type"]))

array([731987.26377803])